modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+from modulus.models.meta import ModelMetaData
+from modulus.models.module import Module
+
+from .modembed import ModEmbedNet
+
+Tensor = torch.Tensor
+
+
+class ScaleShiftMlp(nn.Module):
+    """MLP used to compute the scale and shift parameters of the ModAFNO block
+
+    Parameters
+    ----------
+    in_features : int
+        Input feature size
+    out_features : int
+        Output feature size
+    hidden_features : int, optional
+        Hidden feature size, defaults to 2 * out_features
+    hidden_layers : int, optional
+        Number of hidden layers, defaults to 0
+    activation_fn : nn.Module, optional
+        Activation function, by default nn.GELU
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        hidden_features: Union[int, None] = None,
+        hidden_layers: int = 0,
+        activation_fn: Type[nn.Module] = nn.GELU,
+    ):
+        super().__init__()
+        if hidden_features is None:
+            hidden_features = out_features * 2
+
+        sequence = [nn.Linear(in_features, hidden_features), activation_fn()]
+        for _ in range(hidden_layers):
+            sequence += [nn.Linear(hidden_features, hidden_features), activation_fn()]
+        sequence.append(nn.Linear(hidden_features, out_features * 2))
+        self.net = nn.Sequential(*sequence)
+
+    def forward(self, x: Tensor):
+        (scale, shift) = torch.chunk(self.net(x), 2, dim=1)
+        return (1 + scale, shift)
+
+
+class ModAFNOMlp(AFNOMlp):
+    """Modulated MLP used inside ModAFNO
+
+    Parameters
+    ----------
+    in_features : int
+        Input feature size
+    latent_features : int
+        Latent feature size
+    out_features : int
+        Output feature size
+    activation_fn :  nn.Module, optional
+        Activation function, by default nn.GELU
+    drop : float, optional
+        Drop out rate, by default 0.0
+    scale_shift_kwargs : dict, optional
+        Options to the MLP that computes the scale-shift parameters
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        latent_features: int,
+        out_features: int,
+        mod_features: int,
+        activation_fn: nn.Module = nn.GELU(),
+        drop: float = 0.0,
+        scale_shift_kwargs: Union[dict, None] = None,
+    ):
+        super().__init__(
+            in_features=in_features,
+            latent_features=latent_features,
+            out_features=out_features,
+            activation_fn=activation_fn,
+            drop=drop,
+        )
+        if scale_shift_kwargs is None:
+            scale_shift_kwargs = {}
+        self.scale_shift = ScaleShiftMlp(
+            mod_features, latent_features, **scale_shift_kwargs
+        )
+
+    def forward(self, x: Tensor, mod_embed: Tensor) -> Tensor:
+        (scale, shift) = self.scale_shift(mod_embed)
+
+        scale_shift_shape = (scale.shape[0],) + (1,) * (x.ndim - 2) + (scale.shape[1],)
```

Comment:
Is this ever a variable dim or do we always know the shape will be something like (B, 1, 1, N)

Reply:
Maybe what I had in mind when I wrote this was that (Mod)AFNO can be extended to 1D/3D quite easily and then this would still work... but for the current implementation `x.ndim ==4` always.

Reply:
That makes sense. Was just tracing dims from the model.